In [1]:
from pathlib import Path
import pandas as pd
import numpy as np

print("pandas:", pd.__version__)
print("numpy:", np.__version__)

pandas: 2.3.3
numpy: 2.0.2


In [3]:
from pathlib import Path
import pandas as pd

PROJECT_ROOT = Path.cwd().parent

INPUT_FILE = PROJECT_ROOT / "data" / "student-mat.csv"

df = pd.read_csv(INPUT_FILE, sep=";")

print("Dataset shape:", df.shape)
df.head()

Dataset shape: (395, 33)


,school,sex,age,address,famsize,Pstatus,Medu,Fedu,Mjob,Fjob,...,famrel,freetime,goout,Dalc,Walc,health,absences,G1,G2,G3
0,GP,F,18,U,GT3,A,4,4,at_home,teacher,...,4,3,4,1,1,3,6,5,6,6
1,GP,F,17,U,GT3,T,1,1,at_home,other,...,5,3,3,1,1,3,4,5,5,6
2,GP,F,15,U,LE3,T,1,1,at_home,other,...,4,3,2,2,3,3,10,7,8,10
3,GP,F,15,U,GT3,T,4,2,health,services,...,3,2,2,1,1,5,2,15,14,15
4,GP,F,16,U,GT3,T,3,3,other,other,...,4,3,2,1,2,5,4,6,10,10


In [4]:
cat_cols = df.select_dtypes(include="object").columns

print("Categorical columns:")
print(list(cat_cols))
print("Number of categorical columns:", len(cat_cols))

Categorical columns:
['school', 'sex', 'address', 'famsize', 'Pstatus', 'Mjob', 'Fjob', 'reason', 'guardian', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic']
Number of categorical columns: 17


In [5]:
df_encoded = pd.get_dummies(
    df,
    columns=list(cat_cols),
    drop_first=True
)

print("Original shape:", df.shape)
print("Encoded shape:", df_encoded.shape)

Original shape: (395, 33)
Encoded shape: (395, 42)


# Session 19: Feature Engineering I

## One-Hot Encoding Results

### Original Shape
(395, 33)

### Encoded Shape
(395, 42)

### Summary

The dataset originally contained 395 student records and 33 variables. Seventeen categorical variables were identified and transformed into dummy variables using one-hot encoding. After encoding, the dataset contained 42 columns while preserving all 395 student records.

### Interpretation

The increase from 33 to 42 columns occurred because categorical variables were expanded into binary indicator variables. Variables with multiple categories generated multiple dummy columns. The parameter `drop_first=True` was used to remove one category from each categorical variable, making that category the reference category and reducing redundancy.

The row count remained unchanged because one-hot encoding transforms existing variables rather than adding or removing observations. Each row still represents the same student.

### Recommendation

Verify that:
- The row count remains 395 before and after encoding.
- No object-type columns remain.
- The target variable G3 is unchanged.
- Dummy variables contain only 0 and 1 values.
- No duplicate columns were introduced.

The encoded dataset is now ready for feature scaling and machine-learning model development.

In [6]:
from pathlib import Path

PROJECT_ROOT = Path.cwd().parent

OUTPUT_FILE = (
    PROJECT_ROOT
    / "data"
    / "processed"
    / "student_data_encoded.csv"
)

df_encoded.to_csv(
    OUTPUT_FILE,
    index=False
)

print("Saved:", OUTPUT_FILE)

Saved: /Users/jjun28/Desktop/Github 2026/student-performance-prediction-ml/data/processed/student_data_encoded.csv
